crates/trie/sparse-parallel/src/lower.rs

```diff
@@ -0,0 +1,97 @@
+use crate::SparseSubtrie;
+use reth_trie_common::Nibbles;
+
+/// Tracks the state of the lower subtries.
+///
+/// When a [`crate::ParallelSparseTrie`] is initialized/cleared then its `LowerSparseSubtrie`s are
+/// all blinded, meaning they have no nodes. A blinded `LowerSparseSubtrie` may hold onto a cleared
+/// [`SparseSubtrie`] in order to re-use allocations.
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) enum LowerSparseSubtrie {
+    Blind(Option<Box<SparseSubtrie>>),
+    Revealed(Box<SparseSubtrie>),
+}
+
+impl Default for LowerSparseSubtrie {
+    /// Creates a new blinded subtrie with no allocated storage.
+    fn default() -> Self {
+        Self::Blind(None)
+    }
+}
+
+impl LowerSparseSubtrie {
+    /// Returns a reference to the underlying [`SparseSubtrie`] if this subtrie is revealed.
+    ///
+    /// Returns `None` if the subtrie is blinded (has no nodes).
+    pub(crate) fn as_revealed_ref(&self) -> Option<&SparseSubtrie> {
+        match self {
+            Self::Blind(_) => None,
+            Self::Revealed(subtrie) => Some(subtrie.as_ref()),
+        }
+    }
+
+    /// Returns a mutable reference to the underlying [`SparseSubtrie`] if this subtrie is revealed.
+    ///
+    /// Returns `None` if the subtrie is blinded (has no nodes).
+    pub(crate) fn as_revealed_mut(&mut self) -> Option<&mut SparseSubtrie> {
+        match self {
+            Self::Blind(_) => None,
+            Self::Revealed(subtrie) => Some(subtrie.as_mut()),
+        }
+    }
+
+    /// Reveals the lower [`SparseSubtrie`], transitioning it from the Blinded to the Revealed
+    /// variant, preserving allocations if possible.
+    ///
+    /// The given path is the path of a node which will be set into the [`SparseSubtrie`]'s `nodes`
+    /// map immediately upon being revealed. If the subtrie is blinded, or if its current root path
+    /// is shorter than this one, than this one becomes the new root path of the subtrie.
+    pub(crate) fn reveal(&mut self, path: &Nibbles) {
+        match self {
+            Self::Blind(None) => {
+                *self = Self::Revealed(Box::new(SparseSubtrie::new(*path)));
+            }
+            entry @ Self::Blind(Some(_)) => {
+                let Self::Blind(Some(mut subtrie)) = core::mem::take(entry) else { unreachable!() };
+                subtrie.path = *path;
+                *entry = Self::Revealed(subtrie);
+            }
```

Comment:
I think we can just collapse these two arms into one to avoid doing the `unreachable!()`

```suggestion
            Self::Blind(entry) => {
                if let Some(mut subtrie) = entry.take() {
                    subtrie.path = *path;
                    *self = Self::Revealed(subtrie);
                } else {
                    *self = Self::Revealed(Box::new(SparseSubtrie::new(*path)));
                }
            }
```

Reply:
Nice, done!